### DISCUSSION TITLE
Best optimizations to pass str(n)cmp(...) comparisons with static data ?

### DISCUSSION THREAD

**Discussion by alexandredoyen29** [AFLplusplus/AFLplusplus#2194] [Q&A]
Hi everyone !

Currently, I am working on detection of static data comparisons from program input with functions like strcmp().
My goal is to use AFL++ to infer these static data.

So, I tried different combinations of instrumentation for my target : Cmplog, LAF-Intel and LTO.
But, I found sometimes LAF-Intel is better than Cmplog, and sometimes not in term of inference of static strings.

So, regarding to your experience, when an optimization is better than an other ?
Furthermore, is it possible to combine LAF-Intel with LTO (By using `afl-clang-lto` to compile my target) ? I saw it is possible with Cmplog, but I don't know if it is possible with LAF-Intel...

Have a nice day ;)

**Comment by vanhauser-thc** [comment]
cmplog is better 80% of the time, so of course compcov sometimes is better :) but best is to use both. (but ensure you have a dedicated -M main running that is neither cmplog nor compcov).

LTO works with compcov and cmplog as you can see when reading the output of `afl-cc -h`

**Reply by alexandredoyen29** [reply]
Thank you for your answer !

So, you suggest to have three fuzzing processes :
- The first one, `-M main` that fuzz the "normal" binary compiled with `afl-clang-lto` without cmplog
- The second one, `-S cmplog` that runs AFL++ with `-c binary.cmplog` on the `afl-clang-lto` compiled binary
- The third one, `-S compcov` that runs AFL++ with libcompov as `AFL_PRELOAD=/path/to/libcompcov.so AFL_QEMU_COMPCOV=1 afl-fuzz -Q -S compcov` on the non-instrumented binary

So, at the end, I collect output of the main fuzzing process...

Have a nice day ;)
